[6X backport] Fix ORCA nested SubLink processing during query normalization

hpbee opened the pull request:
This PR is a backport of 
* 3 patches related to improving ORCA behaviour when it normalizes the Query structure with nested SubLinks.
* 1 commit adding testcase for a fixed issue of panic in ORCA which is fixed by one of the patches.
## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

DevChattopadhyay commented:
> Why did you remove queries from subselect.sql ?

There might be some issue while cherry picking the commits from main as the removed cases are not present there. @hpbee can you please fix it.

hpbee commented:
> > Why did you remove queries from subselect.sql ?
> 
> There might be some issue while cherry picking the commits from main as the removed cases are not present there. @hpbee can you please fix it.

Thank you for helping me with understanding the issue @pobbatihari, @DevChattopadhyay. 
Squashed my commit into the original cherry-picked commit with merge issues and rebased my branch.
Also raised a new pipeline.

hpbee commented:
> @hpbee For the first commit `Fix ORCA invalid processing of nested SubLinks with GROUP BY attributes` you added all the outputs of the three commits in subselect.out and subselect_optimizer.out but all the queries are not present in subselect.sql The overall commit looks fine but the individual commits may create doubt for someone seeing the change. Can you please correct it.

Fixed that now @DevChattopadhyay

pobbatihari reviewed (approved):
LGTM.  Make sure that you execute the pipeline before proceeding with the merge.

DevChattopadhyay reviewed (approved):
+1

nimish350 reviewed (approved):
LGTM +1
```
gp_ns_ddl_prod=#  explain                                                                                                                                                          with t0 AS (                                                                                                                                                                     SELECT                                                                                                                                                                             ROW_TO_JSON((SELECT x FROM (SELECT max(t.b)) x))                                                                                                                                AS c                                                                                                                                                                         FROM r                                                                                                                                                                              JOIN s ON true                                                                                                                                                                  JOIN t ON  true                                                                                                                                                            )                                                                                                                                                                             SELECT c FROM t0; 
INFO:  GPORCA failed to produce a plan, falling back to planner
DETAIL:  Feature not supported: Row-type variable
                                                        QUERY PLAN                                                        
--------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=22359949276248.57..22359949276248.60 rows=1 width=32)
   ->  Gather Motion 3:1  (slice3; segments: 3)  (cost=22359949276248.50..22359949276248.55 rows=1 width=4)
         ->  Aggregate  (cost=22359949276248.50..22359949276248.51 rows=1 width=4)
               ->  Nested Loop  (cost=20000000000.00..20764255823748.50 rows=212759127000000 width=4)
                     ->  Nested Loop  (cost=10000000000.00..10240935767.25 rows=2471070000 width=4)
                           ->  Seq Scan on t  (cost=0.00..961.00 rows=28700 width=4)
                           ->  Materialize  (cost=0.00..5696.50 rows=86100 width=0)
                                 ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..4405.00 rows=86100 width=0)
                                       ->  Seq Scan on r  (cost=0.00..961.00 rows=28700 width=0)
                     ->  Materialize  (cost=0.00..5696.50 rows=86100 width=0)
                           ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=0.00..4405.00 rows=86100 width=0)
                                 ->  Seq Scan on s  (cost=0.00..961.00 rows=28700 width=0)
   SubPlan 1  (slice0)
     ->  Subquery Scan on x  (cost=0.00..0.02 rows=1 width=24)
           ->  Result  (cost=0.00..0.01 rows=1 width=0)
 Optimizer: Postgres query optimizer
(16 rows)
```

hpbee commented:
Pushed backported commit of https://github.com/greenplum-db/gpdb/commit/00b3d09aa8d16ad473ec4be943310a38e04a45e5 to this pr.

Stolb27 commented:
Hi there! As far I understand, It's the official position of VMWare to silently ignore the situation with rejection signing CLA for open-source community members without reason and close valid patches (e.g. #16499) without any comments?

kmacoskey commented:
@Stolb27 , more information about reasons for CLA rejections will be coming soon. This looks to be a backport of the changes already accepted previously on the main branch. The authorship in the commits is the same.
